File: heads/python/src/main/python/hydra/lib/chars.py
"""Python implementations of hydra.lib.chars primitives."""

from __future__ import annotations


def is_alpha_num(value: int) -> bool:
    """Check whether a character is alphanumeric."""
    return chr(value).isalnum()


def is_lower(value: int) -> bool:
    """Check whether a character is lowercase."""
    return chr(value).islower()


def is_space(value: int) -> bool:
    """Check whether a character is a whitespace character."""
    return chr(value).isspace()


def is_upper(value: int) -> bool:
    """Check whether a character is uppercase."""
    return chr(value).isupper()


def to_lower(value: int) -> int:
    """Convert a character to lowercase."""
    return ord(chr(value).lower())


def to_upper(value: int) -> int:
    """Convert a character to uppercase."""
    return ord(chr(value).upper())

File: heads/typescript/src/main/typescript/hydra/lib/chars.ts
// Hand-written runtime: hydra.lib.chars primitives.
//
// Characters are represented as Unicode code points (`number`), matching
// Hydra's representation.

export const isAlpha = (c: number): boolean => {
  const ch = String.fromCodePoint(c);
  return /\p{Letter}/u.test(ch);
};

export const isDigit = (c: number): boolean => c >= 48 && c <= 57;
export const isAlnum = (c: number): boolean => isAlpha(c) || isDigit(c);
export const isAlphaNum = isAlnum;
export const isSpace = (c: number): boolean => /\s/.test(String.fromCodePoint(c));
export const isUpper = (c: number): boolean => c >= 65 && c <= 90;
export const isLower = (c: number): boolean => c >= 97 && c <= 122;

export const toUpper = (c: number): number => String.fromCodePoint(c).toUpperCase().codePointAt(0)!;
export const toLower = (c: number): number => String.fromCodePoint(c).toLowerCase().codePointAt(0)!;
